### PR TITLE
[#221] Refactor Validation module

### LINKED ISSUE #221: Refactor Validation module
- [ ] Examples are too hard to follow.
- [ ] Apply our style guide

### PR DESCRIPTION
**Pull request by chshersh**
Resolves #221

I've also added `Monad` instance with custom type errors for `Validation` because why not if we already added a lot of such instances.

### CODE REVIEW

File: src/Relude/Extra/Validation.hs [hunk 1]
```diff
@@ -1,51 +1,78 @@
...
 
-'Validation' is a monoidal sibling to 'Either'. Use 'Validation' on operations that
-might fail with multiple errors that you want to preserve.
--}
+'Validation' is a monoidal sibling to 'Either' but 'Validation' doesn't have a
+'Monad' instance. 'Validation' allows to accumulate all errors instead of
+short-circuting on the first error so you can display all possible errors at
+once. Common use-cases include:
+
+1. Validating each input a form with multiple inputs.
```
Comment (vrom911) [inline src/Relude/Extra/Validation.hs]: "a form" seems to be out of the context

File: src/Relude/Extra/Validation.hs [hunk 2]
```diff
@@ -1,51 +1,78 @@
...
 
+Instances of different standard typeclasses provide various semantics:
+
+1. 'Functor': change the type inside 'Success'.
+2. 'Bifunctor': change both 'Failure' and 'Success'.
+3. 'Applicative': apply function to values inside 'Success' and accumulate
+   errors inside 'Failure'.
+4. 'Semigroup': accumulate both 'Failure' and 'Success' with '<>'.
+5. 'Monoid': 'Success' that shores 'mempty'.
+6. 'Alternative': return first 'Success' or accumulate all errors inside 'Failure'.
```
Comment (vrom911) [inline src/Relude/Extra/Validation.hs]: :fire:

File: src/Relude/Extra/Validation.hs (hunk 1 above)
Comment (chshersh) [reply src/Relude/Extra/Validation.hs]: Ah, yeah, probably I missed a few words...
